src/parser/toplevel.lisp

```diff
@@ -573,32 +666,85 @@ consume all attributes"))))
                  :message "Malformed package declaration"
                  :primary-note "missing package name")))
 
-  ;; Package declarations cannot contain more than two forms
-  (when (cst:consp (cst:rest (cst:rest form)))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:first (cst:rest (cst:rest form))))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "unexpected forms")))
-
-  (unless (identifierp (cst:raw (cst:second form)))
-    (error 'parse-error
-           :err (coalton-error
-                 :span (cst:source (cst:second form))
-                 :file file
-                 :message "Malformed package declaration"
-                 :primary-note "package name must be a symbol")))
-
-  (let* ((package-name (symbol-name (cst:raw (cst:second form))))
-
-         (package (find-package package-name)))
-
-    (unless package
-      (setf package (make-package package-name :use '("COALTON" "COALTON-PRELUDE"))))
-
-    package))
-
+  ;; Any remaining forms are import or export clauses
+  (let ((name-form (cst:nth 1 form)))
+    (unless name-form
+      (error 'parse-error
+             :err (coalton-error
+                   :span (cst:source name-form)
+                   :file file
+                   :message "Malformed package declaration"
+                   :primary-note "missing package name")))
+    (unless (identifierp (cst:raw name-form))
+      (error 'parse-error
+             :err (coalton-error
+                   :span (cst:source name-form)
+                   :file file
+                   :message "Malformed package declaration"
+                   :primary-note "package name must be a symbol")))
+    (let ((package (make-toplevel-package (symbol-name (cst:raw name-form)))))
+      (do-forms (lambda (form)
+                  (parse-package-clause package form file))
+        (cst:rest (cst:rest form)))
+      package)))
+
+
+(defun read-toplevel-package (stream file)
+  "Read a (package) form"
+  (let ((*package* (find-package 'coalton-impl/parser/read)))
+    (multiple-value-bind (form presentp)
+        (maybe-read-form stream *coalton-eclector-client*)
+      (unless presentp
+        (error 'parse-error
+               :err (coalton-error
+                     :span (cons (- (file-position stream) 2)
+                                 (- (file-position stream) 1))
+                     :file file
+                     :message "Unexpected EOF"
+                     :primary-note "missing package form")))
+      (parse-package form file))))
+
+(defun lisp-package (toplevel-package file)
+  (with-slots (name import export) toplevel-package
+    (let ((package (uiop:ensure-package name)))
+      (use-package (find-package "COALTON") package)
```

Comment:
clarified the guarantees, and created doc/package.md

I'm inclined not to support 'midway' packages for now, since they are relatively uncommon, can be supported by keeping those source files as lisp+coalton-toplevel, and they may place new requirements on parsing and code generation.